app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java

```diff
@@ -1,102 +1,213 @@
+            startActivityForResult(i, REQUEST_EXPORT_PATH);
+            return true;
+        });
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        if (DEBUG) {
+            Log.d(TAG, "onActivityResult() called with: requestCode = [" + requestCode + "], resultCode = [" + resultCode + "], data = [" + data + "]");
+        }
+
+        if ((requestCode == REQUEST_IMPORT_PATH || requestCode == REQUEST_EXPORT_PATH)
+                && resultCode == Activity.RESULT_OK) {
+            String path = data.getData().getPath();
+            if(requestCode == REQUEST_EXPORT_PATH) {
+                exportDatabase(path + "/NewPipeData.zip");
+            } else {
+                importDatabase(path);
+            }
+        }
+    }
+
+    private void exportDatabase(String path) {
+        try {
+            ZipOutputStream outZip = new ZipOutputStream(
+                    new BufferedOutputStream(
+                            new FileOutputStream(path)));
+            final String homeDir = getActivity().getApplicationInfo().dataDir;
+            ZipHelper.addFileToZip(outZip, newpipe_db.getPath(), "newpipe.db");
+            ZipHelper.addFileToZip(outZip, newpipe_db_journal.getPath(), "newpipe.db-journal");
+
+            outZip.close();
+
+            Toast.makeText(getContext(), R.string.export_complete_toast, Toast.LENGTH_SHORT)
+                    .show();
+        } catch (Exception e) {
+            onError(e);
+        }
+    }
+
+    private void importDatabase(String filePath) {
+        // check if file is supported
+        ZipFile zipFile = null;
+        try {
+            zipFile = new ZipFile(filePath);
+        } catch (IOException ioe) {
+            Toast.makeText(getContext(), R.string.no_valid_zip_file, Toast.LENGTH_SHORT)
+                    .show();
+            return;
+        } finally {
+            try {
+                zipFile.close();
+            } catch (Exception e){}
+        }
+
+        try {
+            ZipInputStream zipIn = new ZipInputStream(
+                    new BufferedInputStream(
+                            new FileInputStream(filePath)));
+
+            if (!databasesDir.exists() && !databasesDir.mkdir()) {
+                throw new Exception("Could not create databases dir");
+            }
+
+            if(!(ZipHelper.extractFileFromZip(zipIn, newpipe_db.getPath(), "newpipe.db")
+                && ZipHelper.extractFileFromZip(zipIn, newpipe_db_journal.getPath(), "newpipe.db-journal"))) {
+               Toast.makeText(getContext(), R.string.could_not_import_all_files, Toast.LENGTH_LONG)
+                       .show();
+            }
+
+            zipIn.close();
+
+            // restart app to properly load db
+            //App.restart(getContext());
+            System.exit(0);
```

Comment:
This seems to work well, but is this a good practice to circumvent android lifecycle though? I'm not even sure why this works here, since `System.exit(0)` is suppose to pop off the top level settings activity after restarting.

Reply:
Well I'm not sure. The way it should be done according to Stackoverflow is to make the app start using a pending intend with a timer, and after setting this up makeing the application quit.
However if I do it with pending intent the app will restart twice.
I'll check weather this is ok.

Reply:
According to [stackoverflow](https://stackoverflow.com/questions/18292016/difference-between-finish-and-system-exit0?answertab=votes#tab-top) `System.exit()` will close the runtime, but emediatly restart the previous activity.